StreamingKit/StreamingKit/STKHTTPDataSource.m

```diff
@@ -605,4 +620,142 @@ -(BOOL) supportsSeek
+                
+                NSArray *lineItems = [line componentsSeparatedByString:@":"];
+                if (lineItems.count > 1) {
+                    if ([lineItems[0] caseInsensitiveCompare:@"icy-metaint"] == NSOrderedSame) {
+                        metaDataInterval = [lineItems[1] intValue];
+                    } else if ([lineItems[0] caseInsensitiveCompare:@"content-type"] == NSOrderedSame) {
+                        AudioFileTypeID idFromMime = [STKHTTPDataSource audioFileTypeHintFromMimeType:lineItems[1]];
+                        if (idFromMime != 0) {
+                            audioFileTypeHint = idFromMime;
+                        }
+                    }
+                }
+                
+                // this is the end of a line, the new line starts in 2
+                lineStart = streamStart + 2;
+                
+                if (c[2] == '\r' && c[3] == '\n') {
+                    foundIcyEnd = YES;
+                    metaDataPresent = YES;
+                    streamStart += 4; // skip double new line
+                    break;
+                }
+            }
+        }
+    }
+    
+    if (metaDataPresent == YES) {
+        int audioDataByteCount = 0;
+        
+        for (int i = streamStart; i < length; ++i) {
+            // is this a metadata byte?
+            if (metaDataBytesRemaining > 0) {
+                
+                [metaDataString appendFormat:@"%c", buffer[i]];
+                
+                if (--metaDataBytesRemaining == 0) {
+                    dataBytesRead = 0;
+                    
+                    NSDictionary *metadata = [self dictionaryFromMetadata:metaDataString];
+                    [self.delegate dataSource:self didReadMetadata:metadata whileReadingAtOffset:audioDataByteCount];
+                }
+                
+                continue;
+            }
+            
+            // is this the interval byte?
+            if (metaDataInterval > 0 && dataBytesRead == metaDataInterval) {
+                metaDataBytesRemaining = buffer[i] * 16;
+                
+                metaDataString.string = @"";
+                
+                if (metaDataBytesRemaining == 0) {
+                    dataBytesRead = 0;
+                }
+                
+                continue;
+            }
+            
+            // this is a data byte
+            ++dataBytesRead;
+            
+            // overwrite beginning of the buffer with the real audio data
+            // we don't need those bytes any more, since we already examined them
+            buffer[audioDataByteCount++] = buffer[i];
+        }
+        
+        return audioDataByteCount;
+        
+    } else if (foundIcyStart == YES) { // still parsing icy response
+        
+        return 0;
+        
+    } else { // no meta data in stream
+        
+        return length;
+        
+    }
+}
+
+-(NSDictionary *) dictionaryFromMetadata:(NSString *)metaData
```

Comment:
We may want to handle escaped characters and also handle values that are in quotes (key='value') by stripping out the quotes